test-harness/src/test/groovy/com/netflix/counductor/integration/test/DoWhileSpec.groovy

```diff
@@ -0,0 +1,367 @@
+        workflow
+        workflow.status == Workflow.WorkflowStatus.RUNNING
+        workflow.tasks.size() == 6
+
+        when: "Polling for second task"
+        task = workflowExecutionService.getLastPollTask("integration_task_1", "integration.test.worker", null)
+
+        then: "Verify that the expected task is polled"
+        task
+        task.getReferenceTaskName().endsWith(TaskUtils.getLoopOverTaskRefNameSuffix(task.getIteration()))
+
+        when: "The task is completed"
+        task.status = Task.Status.COMPLETED
+        workflowExecutionService.updateTask(task)
+        workflow = workflowExecutionService.getExecutionStatus(workflowInstanceId, true)
+
+        then: "Verify that workflow is running"
+        workflow
+        workflow.status == Workflow.WorkflowStatus.RUNNING
+        workflow.tasks.size() == 6
+
+        when: "Polling for third task"
+        task = workflowExecutionService.getLastPollTask("integration_task_2", "integration.test.worker", null)
+
+        then: "Verify that the expected task is polled"
+        task
+        task.getReferenceTaskName().endsWith(TaskUtils.getLoopOverTaskRefNameSuffix(task.getIteration()))
+
+        when: "The task is completed"
+        task.status = Task.Status.COMPLETED
+        workflowExecutionService.updateTask(task)
+        workflow = workflowExecutionService.getExecutionStatus(workflowInstanceId, true)
+
+        then: "Verify that workflow is completed"
+        workflow
+        workflow.status == Workflow.WorkflowStatus.COMPLETED
+        workflow.tasks.size() == 6
+    }
+
+    def "Test workflow with multiple Do While tasks with multiple iterations"() {
+        given: "Number of iterations of the first loop is set to 2 and second loop is set to 1"
+        def workflowInput = new HashMap()
+        workflowInput['loop'] = 2
+        workflowInput['loop2'] = 1
+
+        when: "The workflow is started"
+        def workflowInstanceId = workflowExecutor.startWorkflow("Do_While_Multiple", 1, "looptest", workflowInput, null, null)
+
+        and: "The workflow is retrieved"
+        def workflow = workflowExecutionService.getExecutionStatus(workflowInstanceId, true)
+
+        then: "Verify that workflow is running"
+        workflow
+        workflow.status == Workflow.WorkflowStatus.RUNNING
+        workflow.tasks.size() == 2
+
+        when: "Polling for first task"
+        def task = workflowExecutionService.getLastPollTask("integration_task_0", "integration.test.worker", null)
+
+        then: "Verify that the expected task is polled"
+        task
+        task.getReferenceTaskName().endsWith(TaskUtils.getLoopOverTaskRefNameSuffix(task.getIteration()))
+
+        when: "The task is completed"
+        task.status = Task.Status.COMPLETED
+        workflowExecutionService.updateTask(task)
+        workflow = workflowExecutionService.getExecutionStatus(workflowInstanceId, true)
+
+        then: "Verify that workflow is running"
+        workflow
```

Comment:
nit: getting the `workflow` state in the `when` can be moved to `then` and assigning of the variable can be skipped by leveraging the `with`: 
```
then:
with(workflowExecutionService.getExecutionStatus(workflowInstanceId, true)){
  status == Workflow.WorkflowStatus.RUNNING
  tasks.size() == 6
  tasks[1].taskType == 'integration_task_0'
  tasks[1].status == Task.Status.COMPLETED
}
```

Reply:
refactored